virocon/distributions.py

```diff
@@ -1494,3 +1495,143 @@ def _fit_mle(self, sample):
+
+    scipy_dist: sts.rv_continuous
+    scipy_dist_name: str
+
+    def __init__(self, *args, **kwargs):
+        try:
+            self.scipy_dist = getattr(sts, self.scipy_dist_name)
+        except AttributeError:
+            pass
+        try:
+            self.scipy_dist
+        except AttributeError:
+            raise TypeError(
+                f"Can't instantiate abstract class {self.__class__.__name__}. "
+                "Overwrite either scipy_dist or scipy_dist_name in subclass."
+            )
+        self.scipy_dist_name = self.scipy_dist.name
+        self._param_names = self._list_scipy_parameters(self.scipy_dist)
+        self._set_default_parameter_values()
+
+        # read parameter from args: (shape(s), loc, scale)
+        assert len(args) <= len(self._param_names)
+        for arg, par_name in zip(args, self._param_names):
+            setattr(self, par_name, arg)
+
+        assert len(kwargs) <= 2 * len(self._param_names)
+        for key, arg in kwargs.items():
+            if key in self._param_names:
+                setattr(self, key, arg)
+            elif key.startswith("f_") and key[2:] in self._param_names:
+                setattr(self, key, arg)
+                setattr(self, key[2:], arg)
+            else:
+                raise TypeError(
+                    f"{self.__class__.__name__} got an unexpected keyword argument '{key}'"
+                )
+
+    #  https://stackoverflow.com/a/53640468
+    @staticmethod
+    def _list_scipy_parameters(distribution):
+        """List parameters for scipy.stats.distribution.
+        # Arguments
+            distribution: a string or scipy.stats distribution object.
+        # Returns
+            A list of distribution parameter strings.
+        """
+        if isinstance(distribution, str):
+            distribution = getattr(sts, distribution)
+        if distribution.shapes:
+            parameters = [name.strip() for name in distribution.shapes.split(",")]
+        else:
+            parameters = []
+        # assert distribution is continous
+        parameters += ["loc", "scale"]
+        return parameters
+
+    def _set_default_parameter_values(self):
+        for par_name in self._param_names:
+            if par_name == "loc":
+                setattr(self, par_name, 0)
+            else:
+                setattr(self, par_name, 1)
+            setattr(self, f"f_{par_name}", None)
+
+    @property
+    def parameters(self):
+        return {par_name: getattr(self, par_name) for par_name in self._param_names}
+
+    def _get_scipy_parameters(self, *args, **kwargs):
+        args_with_default = list(self.parameters.values())
+        for i, arg in enumerate(args):
+            if arg is not None:
+                args_with_default[i] = arg
+
+        for key, arg in kwargs.items():
+            try:
+                idx = self._param_names.index(key)
+            except ValueError:
+                # unknown parameter name
+                assert False
```

Comment:
What is the purpose of "assert False"? I know 'assert' only from tests.

Reply:
Removed